docs/common-sdk-design.md

```diff
@@ -0,0 +1,351 @@
+
+## Problem analysis
+There is nothing fundamentally wrong with current PlayFab SDKs and the SDK generation process. They are technically designed and implemented well, relatively simple, efficient and address most of our today's customers' needs. The simplicity of current implementation - while being a powerful benefit - also has a cost. There are certain limitations that may be standing in a way of even greater adoption of the platform by customers with specific needs or extending business opportunities provided by the platform:
+* Monolithic, one-for-all formation of current SDKs results in:
+    * Limited configurability (ability to select which components are needed or should be excluded from the SDK)
+    * Limited customization options (ability to easily replace standard components with custom ones in the SDK)
+    * Limited extensibility (ability to easily add new external functionality to the SDK)
+* No possibility to build a custom SDK
+* No easy and configurable way to:
+    * Use an alternative service platform (e.g. a different data ingestion system)
+    * Use a different network transport/protocol (unless already built in the SDK)
+    * Use different elements of data processing stack on the client side (e.g. specific data formatting, serialization, compression, encryption)
+    * Use alternative client-side API
+*  No diversity in available solutions (beside a single SDK (per client platform) and REST API) that would allow a wider spectrum of integration options with PlayFab services (think of a "marketplace" of various solutions created by PlayFab and third parties, that would allow users to choose the breadth and the depth of interaction with PlayFab platform)
+
+## Solution
+Many limitations listed above can be addressed by a set of improvements in SDKs themselves and in the process of their generation:
+
+* SDKs:
+    * Modular composition (SDK consists of smaller building blocks - _plugins_)
+    * Dynamic formation (plugins can be removed, added, replaced) based on SDK configuration
+    * Custom plugins (created by users) are allowed
+    * Plugins can interact with other plugins
+
+* SDK generation process:
+    * SDKs are not directly downloaded from a single designated git repo as one-for-all content anymore. Instead, they become customer-specific content that can be personally designed from plugins, configured, packaged and then downloaded by user.
+    * The SDK generation consists of two separate processes:
+        * _Plugin Generator_ (similar to the current SDK Generator but generates plugins instead of whole SDKs. Each plugin is stored in its own git repo (similarly as a whole SDK now))
+        * _Common SDK Builder_ (with user's SDK configuration as an input it assembles a custom SDK from plugins, generates some "gluing" code that allows plugins to interoperate, packages it up as a complete SDK with source code and project files and prepares it as downloadable content for user)
```

Comment:
We can totally ship binaries instead of (or in addition to) the source code, either for a particular plugin or even the whole SDK. This design will not restrict us from doing that. As a side note, all our current monolithic SDK are shipped for all platforms as source code only. They are git repos with source code.

---
In reply to: [194566921](https://github.com/PlayFab/SDKGenerator/pull/134#discussion_r194566921) [](ancestors = 194566921)